org.eclipse.xtext/src/org/eclipse/xtext/nodemodel/impl/NodeModelBuilder.java

```diff
@@ -32,54 +30,58 @@
  */
 public class NodeModelBuilder {
 
-	private static class ArrayInterner<T> implements Interner<T[]> {
+	private static class ArrayInterner {
 
-		private static class ArrayAsList<T> extends AbstractList<T> implements RandomAccess {
-			final T[] array;
-			private int hashCode = -1;
+		private static class InternKey {
+			final EObject myGrammarElement;
+			final Object childGrammarElement;
+			
+			int hashCode = -1;
 
-			ArrayAsList(/* @Nullable */ T[] array) {
-				this.array = array;
+			InternKey(EObject myGrammarElement, Object childGrammarElement) {
+				this.myGrammarElement = myGrammarElement;
+				this.childGrammarElement = childGrammarElement;
 			}
-
-			@Override public int size() {
-				return array.length;
-			}
-
-			@Override /* @Nullable */ public T get(int index) {
-				return array[index];
+			
+			EObject[] joinElements() {
+				if(childGrammarElement instanceof EObject) {
+					return new EObject[] {myGrammarElement, (EObject) childGrammarElement};
+				} else {
+					return ObjectArrays.concat(myGrammarElement, (EObject[]) childGrammarElement);
+				}
 			}
 
-			@Override public int hashCode() {
+			@Override
+			public int hashCode() {
 				if (hashCode == -1)
-					hashCode = Arrays.hashCode(array);
+					hashCode = Objects.hash(myGrammarElement, childGrammarElement);
 				return hashCode;
 			}
 
-			@Override public boolean equals(/* @Nullable */ Object o) {
-				if (this == o)
-					return true;
-				return o instanceof ArrayAsList<?> && Arrays.equals(array, ((ArrayAsList<?>) o).array);
+			@Override
+			public boolean equals(Object o) {
+				if(o instanceof InternKey) {
+					InternKey interned = (InternKey) o;
+					return Objects.equals(this.myGrammarElement, interned.myGrammarElement) 
+							&& Objects.equals(this.childGrammarElement, interned.childGrammarElement);
+				}
+				
+				return false;
 			}
 		}
 
-		private Map<ArrayAsList<T>, T[]> map = Maps.newHashMap();
-
-		@Override
-		public /* @Nullable */ T[] intern(/* @Nullable */ T[] sample) {
-			ArrayAsList<T> key = new ArrayAsList<T>(sample);
-			T[] canonical = map.get(key);
-			if (canonical == null) {
-				canonical = sample;
-				map.put(key, canonical);
-			}
-	        return canonical;
+		Map<InternKey, EObject[]> interningMap = Maps.newHashMap();
```

Comment:
This causes a leak. The ArrayInterner instance holds a strong referenced map.
You may get an inspiration of weak interning from EMF. Look at org.eclipse.emf.common.util.WeakInterningHashSet and its usage.

Reply:
No, this map is disposed together with the NodeModelBuilder after the resource was parsed.